src/codegen.cpp

```diff
@@ -2279,6 +2279,9 @@ void walk_function_params(CodeGen *g, ZigType *fn_type, FnWalk *fn_walk) {
                 if ((param_type->id == ZigTypeIdPointer && param_type->data.pointer.is_const) || is_byval) {
                     addLLVMArgAttr(llvm_fn, (unsigned)gen_index, "readonly");
                 }
+                if (param_type->id == ZigTypeIdPointer && param_type->data.pointer.explicit_alignment != 0) {
```

Comment:
We can do even better than `param_type->id == ZigTypeIdPointer`, handling `?*align(n) T` (optional pointers) as well. I believe you want `get_codegen_ptr_type(param_type) != nullptr`. The function you already chose, `get_ptr_align`, supports any type for which `get_codegen_ptr_type` returns non-null.

When you do this, accessing `data.pointer.explicit_alignment` however will only work for `ZigTypeIdPointer`, so there are a couple options:
 * Put the `align <n>` attribute on all pointer parameters. I'm sure this is fine.
 * Introduce a new function in analyze.cpp, `get_ptr_align_if_not_abi_aligned` which returns the alignment, or 0 if ABI aligned, in which case you can avoid adding the attribute.

Reply:
> Put the align <n> attribute on all pointer parameters. I'm sure this is fine.

I went for the easiest solution, after all llvm is going to compute the alignment anyway.